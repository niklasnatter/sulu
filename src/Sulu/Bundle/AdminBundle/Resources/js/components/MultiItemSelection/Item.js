// @flow
import React from 'react';
import type {Node} from 'react';
import {SortableHandle} from 'react-sortable-hoc';
import classNames from 'classnames';
import Icon from '../Icon';
import itemStyles from './item.scss';

const DRAG_ICON = 'su-more';

type Props<T> = {
    children: Node,
    disabled: boolean,
    id: T,
    index: number,
    onEdit?: (id: T) => void,
    onRemove?: (id: T) => void,
    sortable: boolean,
};

export default class Item<T> extends React.PureComponent<Props<T>> {
    static defaultProps = {
        disabled: false,
        sortable: true,
    };

    createDragHandle() {
        const {sortable} = this.props;

        const handle = ({className, children}: Object) => (
            <span className={className}>{children}</span>
        );

        if (!sortable) {
            return handle;
        }

        return SortableHandle(handle);
    }

    handleEdit = () => {
        const {id, onEdit} = this.props;

        if (onEdit) {
            onEdit(id);
        }
    };

    handleRemove = () => {
        const {id, onRemove} = this.props;

        if (onRemove) {
            onRemove(id);
        }
    };

    render() {
        const {
            children,
            disabled,
            index,
            onEdit,
            onRemove,
            sortable,
        } = this.props;

        const DragHandle = this.createDragHandle();

        const itemClass = classNames(
            itemStyles.item,
            {
                [itemStyles.disabled]: disabled,
            }
        );

        const dragHandleClass = classNames(
            itemStyles.dragHandle,
            {
                [itemStyles.sortable]: sortable,
            }
        );

        return (
            <div className={itemClass}>
                <DragHandle className={dragHandleClass}>
                    {sortable && <Icon name={DRAG_ICON} />}
                    <span className={itemStyles.index}>{index}</span>
                </DragHandle>
                <div className={itemStyles.content}>
                    {children}
                </div>
                <div className={itemStyles.buttons}>
                    {onEdit &&
                        <button className={itemStyles.button} onClick={this.handleEdit} type="button">
                            <Icon name="su-pen" />
                        </button>
                    }
                    {onRemove &&
                        <button className={itemStyles.button} onClick={this.handleRemove} type="button">
                            <Icon name="su-trash-alt" />
                        </button>
                    }
                </div>
            </div>
        );
    }
}
