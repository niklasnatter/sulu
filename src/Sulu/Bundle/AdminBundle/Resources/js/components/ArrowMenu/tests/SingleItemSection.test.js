// @flow
import React from 'react';
import {render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SingleItemSection from '../SingleItemSection';
import Item from '../Item';

test('Render ItemSection', () => {
    const {container} = render(
        <SingleItemSection
            icon="fa-home"
            onChange={jest.fn()}
            title="Select your house"
            value={undefined}
        >
            <Item value="villa">Villa</Item>
            <Item value="white_house">White House</Item>
            <Item value="flat">Flat</Item>
        </SingleItemSection>
    );

    expect(container).toMatchSnapshot();
});

test('Render ItemSection with value', () => {
    const {container} = render(
        <SingleItemSection
            icon="fa-home"
            onChange={jest.fn()}
            title="Select your house"
            value="flat"
        >
            <Item value="villa">Villa</Item>
            <Item value="white_house">White House</Item>
            <Item value="flat">Flat</Item>
        </SingleItemSection>
    );

    expect(container).toMatchSnapshot();
});

test('Handle Item click', async() => {
    const handleChange = jest.fn();
    render(
        <SingleItemSection
            icon="fa-home"
            onChange={handleChange}
            title="Select your house"
            value={undefined}
        >
            <Item value="villa">Villa</Item>
            <Item value="white_house">White House</Item>
            <Item value="flat">Flat</Item>
        </SingleItemSection>
    );

    const user = userEvent.setup();
    await user.click(screen.getByText('White House'));

    expect(handleChange).toBeCalledWith('white_house');
});
