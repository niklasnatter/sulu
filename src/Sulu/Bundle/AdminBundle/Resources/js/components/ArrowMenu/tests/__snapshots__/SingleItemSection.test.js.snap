// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ItemSection 1`] = `
<div>
  <div
    class="section"
  >
    <div
      class="title"
    >
      Select your house
    </div>
    <div
      class="children"
    >
      <button
        class="item"
        type="button"
      >
        <span
          class="icon"
        />
        <span>
          Villa
        </span>
      </button>
      <button
        class="item"
        type="button"
      >
        <span
          class="icon"
        />
        <span>
          White House
        </span>
      </button>
      <button
        class="item"
        type="button"
      >
        <span
          class="icon"
        />
        <span>
          Flat
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Render ItemSection with value 1`] = `
<div>
  <div
    class="section"
  >
    <div
      class="title"
    >
      Select your house
    </div>
    <div
      class="children"
    >
      <button
        class="item"
        type="button"
      >
        <span
          class="icon"
        />
        <span>
          Villa
        </span>
      </button>
      <button
        class="item"
        type="button"
      >
        <span
          class="icon"
        />
        <span>
          White House
        </span>
      </button>
      <button
        class="item active"
        type="button"
      >
        <span
          class="icon"
        >
          <span
            aria-label="fa-home"
            class="fa fa-home icon"
          />
        </span>
        <span>
          Flat
        </span>
      </button>
    </div>
  </div>
</div>
`;
